analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs

```diff
@@ -50,27 +61,62 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
             }
         }
 
         private static bool CheckExpression(SyntaxNode node, ExpressionSyntax modulus, SemanticModel semanticModel, out int constantValue) =>
             ExpressionNumericConverter.TryGetConstantIntValue(node, out constantValue)
             && constantValue != 0
             && IsModulus(modulus)
-            && !IsUnsigned(modulus, semanticModel);
+            && !ExpressionIsAlwaysPositive(modulus, semanticModel);
 
         private static bool IsModulus(ExpressionSyntax expression) =>
             expression.RemoveParentheses() is BinaryExpressionSyntax binary
             && binary.IsKind(SyntaxKind.ModuloExpression);
 
-        private static bool IsUnsigned(ExpressionSyntax expression, SemanticModel semantic)
+        private static bool ExpressionIsAlwaysPositive(ExpressionSyntax expression, SemanticModel semantic)
         {
             var type = semantic.GetTypeInfo(expression).Type;
-            return type.IsAny(KnownType.UnsignedIntegers)
-                   || type.Is(KnownType.System_UIntPtr);
+            var isUint = type.IsAny(KnownType.UnsignedIntegers)
+                          || type.Is(KnownType.System_UIntPtr);
+
+            var leftExpression = ((BinaryExpressionSyntax)expression).Left;
+            if (!isUint && CollectionSizePropertyOrMethodNames.Any(x => leftExpression.ToString().Contains(x)))
+            {
+                var symbol = semantic.GetSymbolInfo(((BinaryExpressionSyntax)expression).Left).Symbol;
+                return IsCollectionSizeMethodOrProperty(symbol);
+            }
+            return isUint;
         }
+
+        private static bool IsCollectionSizeMethodOrProperty(ISymbol symbol) =>
+            IsEnumerableCountMethod(symbol)
+            || IsLengthProperty(symbol)
+            || IsCollectionCountProperty(symbol)
+            || IsListCapacityProperty(symbol);
+
+        private static bool IsEnumerableCountMethod(ISymbol symbol) =>
+            (CountName.Equals(symbol.Name) || LongCountName.Equals(symbol.Name))
+            && symbol is IMethodSymbol methodSymbol
+            && methodSymbol.IsExtensionMethod
+            && methodSymbol.ReceiverType != null
+            && methodSymbol.IsExtensionOn(KnownType.System_Collections_Generic_IEnumerable_T);
+
+        private static bool IsLengthProperty(ISymbol symbol) =>
+            (LengthName.Equals(symbol.Name) || LongLengthName.Equals(symbol.Name))
+            && symbol is IPropertySymbol propertySymbol
+            && (propertySymbol.ContainingType.Is(KnownType.System_Array) || propertySymbol.ContainingType.Is(KnownType.System_String));
+
+        private static bool IsCollectionCountProperty(ISymbol symbol) =>
+            CountName.Equals(symbol.Name)
+            && symbol is IPropertySymbol propertySymbol
+            && (propertySymbol.ContainingType.DerivesOrImplements(KnownType.System_Collections_Generic_ICollection_T)
```

Comment:
will `Implements()` suffice, given that you're only checking interfaces? what's the use case for derives? I cannot see it in the UTs

Reply:
I kept also `Derives()` for safety. I think I will keep only `Implements()` to keep it simple for now.
WDYT?

Reply:
Update: After doing a quick test `Derives` is not really needed at all.

Reply:
you can add that test to the UT to document the expected behavior